### PR TITLE
Mouse drag and cell edit fixes for ticket 1064

### PR DESCRIPTION
**Pull request by wdednam**
The immutability of row 1 guaranteed now.

However, a few tests on lines 1288,1289, 1298, 1299 of testModel.cc fail with the following output:

testModel.cc:1288:1: error: Failure in mouseButtons: Expected r2c1 but was 
testModel.cc:1289:1: error: Failure in mouseButtons: Expected r1c1 but was 
testModel.cc:1298:1: error: Failure in mouseButtons: Expected r3c1 but was 
testModel.cc:1299:1: error: Failure in mouseButtons: Expected r2c1 but was 

It appears cells in column 1 cannot be clicked on by the mouse, even in rows beneath the initial conditions (row 1)? I cannot pinpoint the reasons from mouseDown(), starting on line 430 in godleyTableWindow.cc, and I have not modified any code between lines 432 and 472.

---
This change is [Reviewable](https://reviewable.io/reviews/highperformancecoder/minsky/69)

### CODE REVIEW

File: gui-tk/godley.tcl [hunk 1]
```diff
@@ -230,12 +230,12 @@ proc godleyContext {id x y X Y} {
             $id.insertIdx 0
             $id.selectIdx 0
         }
-        if [string length [$id.godleyIcon.table.getCell $r $c]] {
+        if {[string length [$id.godleyIcon.table.getCell $r $c]] && [$id.godleyIcon.table.getCell $r $c]!=[$id.godleyIcon.table.getCell 1 0]} {   # Cannot Cut or Copy cell(1,0). For ticket 1064
```
Comment (highperformancecoder) [inline gui-tk/godley.tcl]: Why not make the condition && ($r!=1 || $c!=0) ? Currently, you are just testing that the source cell is not "Iniitial  condtitions", which is not the same thing.

File: gui-tk/godley.tcl [hunk 2]
```diff
@@ -230,12 +230,12 @@ proc godleyContext {id x y X Y} {
...
             .$id.context add command -label "Cut" -command "$id.cut"
             .$id.context add command -label "Copy" -command "$id.copy"
         }
     }
-    if {![catch {clipboard get -type UTF8_STRING}]} {
+    if {![catch {clipboard get -type UTF8_STRING}] && [$id.godleyIcon.table.getCell $r $c]!=[$id.godleyIcon.table.getCell 1 0]} {   # Cannot Paste into cell(1,0). For ticket 1064
```
Comment (highperformancecoder) [inline gui-tk/godley.tcl]: Ditto

File: gui-tk/godley.tcl (hunk 1 above)
Comment (highperformancecoder) [reply gui-tk/godley.tcl]: Also, there is no problem in allowing the user to copy cell (1,0).

File: model/godleyTableWindow.cc [hunk 3]
```diff
@@ -414,12 +414,15 @@ namespace minsky
         selectedCol>=0 && size_t(selectedCol)<godleyIcon->table.cols())
       {
         auto& str=godleyIcon->table.cell(selectedRow,selectedCol);
-        pango.setMarkup(defang(str));
-        int j=0;
-        if (selectedCol>=int(scrollColStart)) j=selectedCol-scrollColStart+1;
-        x-=colLeftMargin[j]+2;
-        x*=zoomFactor;
-        return x>0 && str.length()>0?pango.posToIdx(x)+1: 0;
+        if (str!=godleyIcon->table.cell(1,0))                           // No text index needed for a cell that is immutable. For ticket 1064
```
Comment (highperformancecoder) [inline model/godleyTableWindow.cc]: Don't you just want tp avoid (selectedRow,selectedCol)!=(1,0)? We can't assume that (1,0) is the only cell with "Initial Conditions" in it.

File: model/godleyTableWindow.cc [hunk 4]
```diff
@@ -463,10 +466,12 @@ namespace minsky
...
         if (selectedRow>=0 && selectedRow<int(godleyIcon->table.rows()) &&
             selectedCol>=0 && selectedCol<int(godleyIcon->table.cols()))
-          {
-            selectIdx=insertIdx = textIdx(x);
-            savedText=godleyIcon->table.cell(selectedRow, selectedCol);
-          }
+           {
+             selectIdx=insertIdx = textIdx(x);
+             auto& str=godleyIcon->table.cell(selectedRow,selectedCol);     
+             if (str!=godleyIcon->table.cell(1,0))                               // Cannot save text in cell(1,0). For ticket 1064
```
Comment (highperformancecoder) [inline model/godleyTableWindow.cc]: Same - you should just test directly on selectedCol, selectedRow.

File: model/godleyTableWindow.cc [hunk 5]
```diff
@@ -487,10 +492,10 @@ namespace minsky
       }
     else if (r>0 && selectedCol==0)
       {
-        if (r!=selectedRow)
+        if (r!=selectedRow && !godleyIcon->table.initialConditionRow(selectedRow) && !godleyIcon->table.initialConditionRow(r))  // Cannot move Intitial Conditions row. For ticket 1064
           godleyIcon->table.moveRow(selectedRow,r-selectedRow);
       }
-    else if ((c!=selectedCol || r!=selectedRow) && c>0 && r>0)
+    else if ((c!=selectedCol || r!=selectedRow) && c>0 && r>0 && !godleyIcon->table.initialConditionRow(selectedRow) && !godleyIcon->table.initialConditionRow(r)) // Cannot swap individual cells in IC row. For ticket 1064
```
Comment (highperformancecoder) [inline model/godleyTableWindow.cc]: This is probably a little too restrictive. Just check that (c,r)!=(0,1) and (selectedCol,selectedRow)!=(0,1)

File: model/godleyTableWindow.cc [hunk 6]
```diff
@@ -553,77 +558,78 @@ namespace minsky
...
-              if (insertIdx>=str.length()) insertIdx=str.length();
-              str.insert(insertIdx,utf8);
-              selectIdx=insertIdx+=utf8.length();
-            }
-          else
-            {
-              switch (utf8[0]) // process control characters
+          {			  	  
+            auto& str=table.cell(selectedRow,selectedCol);
+            if (str!=table.cell(1,0))                                 // Cell (1,0) is off-limits. For ticket 1064
```
Comment (highperformancecoder) [inline model/godleyTableWindow.cc]: Again - just test selectedCol,selectedRow

File: gui-tk/godley.tcl (hunk 1 above)
Comment (wdednam) [reply gui-tk/godley.tcl]: Thanks :) Yes, testing for string "Initial Conditions" produces unexpected behaviour in other cells.

File: gui-tk/godley.tcl (hunk 2 above)
Comment (wdednam) [reply gui-tk/godley.tcl]: Sorted, thanks.

File: model/godleyTableWindow.cc (hunk 3 above)
Comment (wdednam) [reply model/godleyTableWindow.cc]: Very true, using (selectedCol!=0 || selectedRow!=1) instead, thank you.

File: model/godleyTableWindow.cc (hunk 4 above)
Comment (wdednam) [reply model/godleyTableWindow.cc]: Done

File: model/godleyTableWindow.cc (hunk 5 above)
Comment (wdednam) [reply model/godleyTableWindow.cc]: Okay, cells other than (1,0) in the first row can now also be swapped out for a (selectedCol!=0 || selectedRow!=1) condition.

File: model/godleyTableWindow.cc (hunk 6 above)
Comment (wdednam) [reply model/godleyTableWindow.cc]: Also done

File: test/testModel.cc [hunk 7]
```diff
@@ -1285,8 +1285,8 @@ SUITE(GodleyTableWindow)
       CHECK_EQUAL("r2c2",godleyIcon->table.cell(2,2));
       CHECK_EQUAL("r1c2",godleyIcon->table.cell(1,2));
       // Not sure why column 1's contents cannot be tested????
-      CHECK_EQUAL("r2c1",godleyIcon->table.cell(2,1));
-      CHECK_EQUAL("r1c1",godleyIcon->table.cell(1,1));
+      //CHECK_EQUAL("r2c1",godleyIcon->table.cell(2,1));
+      //CHECK_EQUAL("r1c1",godleyIcon->table.cell(1,1));
```
Comment (highperformancecoder) [inline test/testModel.cc]: Column 1 has been moved to column 3 at line 1274. But I think it is enough of a test to test on column 2, like you do, so this code can be removed.

File: test/testModel.cc [hunk 8]
```diff
@@ -1295,8 +1295,8 @@ SUITE(GodleyTableWindow)
       CHECK_EQUAL("r3c2",godleyIcon->table.cell(2,2));
       CHECK_EQUAL("r2c2",godleyIcon->table.cell(3,2));
       // Not sure why column 1's contents cannot be tested????
-      CHECK_EQUAL("r3c1",godleyIcon->table.cell(2,1));
-      CHECK_EQUAL("r2c1",godleyIcon->table.cell(3,1));
+      //CHECK_EQUAL("r3c1",godleyIcon->table.cell(2,1));
+      //CHECK_EQUAL("r2c1",godleyIcon->table.cell(3,1));
```
Comment (highperformancecoder) [inline test/testModel.cc]: Ditto.

File: test/testModel.cc (hunk 7 above)
Comment (wdednam) [reply test/testModel.cc]: So no wonder they appear empty in the comparison, I never defined initial "r*c3" values for column 3. I might not have been surprised to see a mismatch, but was confused by the fact that the comparison was with empty values. Thanks for clearing that up. :)

File: test/testModel.cc (hunk 8 above)
Comment (wdednam) [reply test/testModel.cc]: I've removed lines 1288, 1289, 1298 and 1299 and tidied up the test a bit.

**Pull request merged**